microcosm_flask/audit.py

```diff
@@ -2,80 +2,131 @@
+
+    """
+    def include_header(header):
+        return header.startswith(prefix)
+    return include_header
+
+
 def audit(func):
     """
     Record a Flask route function in the audit log.
 
     Generates a JSON record in the Flask log for every request.
 
     """
-    logger = getLogger("audit")
-
     @wraps(func)
     def wrapper(*args, **kwargs):
-        response = None
-
-        # always include these fields
-        audit_dict = dict(
-            operation=request.endpoint,
-            func=func.__name__,
-            method=request.method,
+        options = AuditOptions(
+            include_request_body=True,
+            include_response_body=True,
+            include_header=include_header_matching_prefix(X_REQUEST),
         )
+        return _audit_request(options, func, *args, **kwargs)
 
-        # include request body on debug (if any)
-        if current_app.debug and request.get_json(force=True, silent=True):
-            audit_dict["request_body"] = request.get_json(force=True)
-
-        try:
-            response = func(*args, **kwargs)
-        except Exception as error:
-            status_code = extract_status_code(error)
-            success = 0 < status_code < 400
-            audit_dict.update(
-                success=success,
-                message=extract_error_message(error)[:2048],
-                context=extract_context(error),
-                stack_trace=None if success else format_exc(limit=10),
-                status_code=status_code,
-            )
-            raise
-        else:
-            body, status_code = parse_response(response)
+    return wrapper
 
-            audit_dict.update(
-                success=True,
-                status_code=status_code,
-            )
 
-            # include response body on debug (if any)
-            if current_app.debug and body and not g.get("hide_body"):
-                try:
-                    audit_dict["response_body"] = loads(body)
-                except (TypeError, ValueError):
-                    # not json
-                    audit_dict["response_body"] = body
+def _audit_request(options, func, *args, **kwargs):
+    """
+    Run a request function under audit.
 
-            return response
-        finally:
-            # always log at INFO; we can't know whether a raised exception
-            # is an error or expected behavior
-            logger.info(audit_dict)
+    """
+    logger = getLogger("audit")
 
-    return wrapper
+    response = None
+
+    # always include these fields
+    audit_dict = dict(
+        operation=request.endpoint,
+        func=func.__name__,
+        method=request.method,
+    )
+
+    # include request body on debug (if any)
+    if all((
+        current_app.debug,
+        options.include_request_body,
+        request.get_json(force=True, silent=True),
+        not g.get("hide_body"),
+    )):
+        audit_dict["request_body"] = request.get_json(force=True)
```

Comment:
Why convert this to a dict?  Isn't this the cause for invalid json being logged? ie. u'